lib/xwax/pitch_kalman.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Kalman filter pitch estimator with sensitivity mode switches
+ *
+ * Model: Constant velocity (x, v) with acceleration process noise
+ *
+ * State:
+ *   x : Position on the unit circle or complex plane (see the scope)
+ *   v : Velocity (pitch)
+ *
+ * Assumption: We steadily advance with a constant velocity
+ *   dx: Usually the displacement is 1/4 of cycle of the sinusoid π/2
+ *       or one zero crossing
+ *
+ * Observation:
+ *   z = dx (new measurement after the interval dt)
+ *
+ * Modes:
+ *   stable  : Low Q and high R for stable playback
+ *   medium  : Medium values for slight pitch changes
+ *   reactive: High Q and low R for high reactivity (scratching)
+ */
+
+#ifndef PITCH_KALMAN_H
+#define PITCH_KALMAN_H
+
+#include <errno.h>
+#include <stdio.h>
+
+/* Struct for the Kalman filter coefficients R and Q */
+
+struct kalman_coeffs {
+    double Q;
+    double R;
+};
+
+#define KALMAN_COEFFS(q_arg, r_arg) \
+    (struct kalman_coeffs) {.Q = (q_arg), .R = (r_arg)}
+
+struct pitch_kalman {
+    /*
+     * NOTE: In discrete time dt is usually denoted as Ts = 1/Fs,
+     * but xwax uses the continuous notation.
+     */
+
+    double dt; /* Sampling interval: (s) */
+
+    double Xk[2]; /* Position and velocity state space */
+
+    /* Covariance P (2x2, symmetric) */
+
+    double P[2][2];
+
+    /* Tresholds of the innovation quantity for the mode switches */
+
+    double reactive_threshold, medium_threshold;
+
+    /* Currently used coefficients*/
+
+    struct kalman_coeffs* coeffs;
+
+    /* Stable, medium reactive coefficients for the mode switch */
+
+    struct kalman_coeffs stable;
+    struct kalman_coeffs medium;
+    struct kalman_coeffs reactive;
+};
+
+void pitch_kalman_init(struct pitch_kalman* p, double dt, struct kalman_coeffs stable,
+        struct kalman_coeffs medium, struct kalman_coeffs reactive,
+        double medium_threshold, double reactive_threshold);
+void pitch_kalman_update(struct pitch_kalman* p, double dx);
+
+/*
+ * Retune noise sensitivity without resetting state
+ */
+
+static inline void kalman_tune_sensitivity(struct pitch_kalman* p, struct kalman_coeffs* coeffs)
+{
+    if (!p || !coeffs) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    p->coeffs = coeffs;
+}
+
+
+/*
+ * Get the current pitch (velocity estimate)
+ */
+
+static inline double pitch_kalman_current(const struct pitch_kalman* p)
+{
+    if (!p) {
+        errno = EINVAL;
+        perror(__func__);
+        return 0.0;
+    }
+
+    /* Return the velocity Xk[v] */
```

Comment:
Xk[v] / tc->speed is the rate ratio which is 1 at unity. tc->speed is also 1 for a 33 r/min vinyl. 

See: 
https://github.com/mixxxdj/mixxx/blob/ba4a0f9e11aefe0bc9f1ce6328da26ed89840adf/src/vinylcontrol/vinylcontrolxwax.cpp#L144

```suggestion
    /* Return the velocity Xk[v] relative to 33.3 r/min */
```

Reply:
Improved the comment